[자동차 경주 구현] 메다 미션 제출합니다.

medAndro opened the pull request:
안녕하세요 리뷰어님! 

이번에 저희 페어는 의도적으로 MVC를 사용하지 않고 코드를 작성하여 MVC모델을 사용하지 않았을 때의 발생하는 문제점을 파악하고 이후  MVC모델을 적용하는 리팩터링을 진행하는 방향으로 코드를 구현 하였고 테스트 코드를 작성하기 더 수월해졌다고 느꼈습니다.

### 궁금한점
- 테스트코드에서 현재 확률이 60%~60.01% 사이인 경우를 테스트하고 있는데 아주 낮은 확률로 테스트가 실패할 것이라 생각되는데. 랜덤값이 포함되는 테스트코드를 작성할 때 현재의 방법이 괜찮은것인지 아니면 더 나은 방법이 있는지 궁금합니다.

- 현재 입출력을 받을 때 View클래스를 만들지 않고 있는데, 구현한다면 단순히 아래 코드블럭처럼 readln()과 println()을 감싸게 되는 불필요한 구조가 될 것 같아서 생략하였습니다. 이러한 접근으로 View 클래스를 생략하는게 좋을지 궁금합니다.
```kt
class View {
    fun readLine(): String {
        return readln()
    }
    fun printLine(message: String) {
        println(message)
    }
}
```

medAndro committed:
docs: 구현 기능 목록 작성

medAndro committed:
feat: 프로젝트 기본 구조 작성

- 출력할 문자열들을 관리하는 Messages enum 추가
- 자동차 이름과 경주 횟수를 입력받는 로직 추가

medAndro committed:
feat: 입력값에 대한 예외 처리 구현

- InputValidator 클레스에 자동차 이름과 시도 횟수 예외를 테스트하는 메서드 구현

oungsi2000 committed:
feat(Car.kt) : Car 객체 구현

oungsi2000 committed:
feat:전체 기능 완성

oungsi2000 committed:
fix:랜덤값 버그 수정

oungsi2000 committed:
refactor:전체 리팩토링

oungsi2000 committed:
feat:테스트 코드 작성

medAndro committed:
refactor: 코드 컨벤션 및 테스트코드 호환성 수정

- klint 컨벤션 적용
- 테스트코드에서 줄 끝 문자의 윈도우/맥 호환을 위한 확장함수 추가

medAndro committed:
feat: 문자열 예외 테스트 케이스 추가

oungsi2000 committed:
merge test

medAndro committed:
refactor: MVC 구조로 변경

-  Application.kt 내 서비스 로직을 inputValidatorService으로 분리

oungsi2000 committed:
docs:readme 작성

oungsi2000 committed:
docs:readme 수정

laco-dev commented on src/test/kotlin/ApplicationTest.kt:
```diff
@@ -0,0 +1,156 @@
+        )
+    }
+
+    @Test
+    @DisplayName("랜덤 값 추출 테스트, 확률이 60%~60.01% 사이여야 한다")
+    fun t6() {
+        val random = Random(RANDOM_SEED)
+        var count = 0
+        val osName = System.getProperty("os.name")
+        assertThat(osName).contains("MAc")
+        repeat(100000000) { if (random.nextInt(0, 10) >= 4) count++ }
+        assertThat(count)
+            .isGreaterThan(60000000)
+            .isLessThan(60010000)
+    }
```
> 테스트코드에서 현재 확률이 60%~60.01% 사이인 경우를 테스트하고 있는데 아주 낮은 확률로 테스트가 실패할 것이라 생각되는데. 랜덤값이 포함되는 테스트코드를 작성할 때 현재의 방법이 괜찮은것인지 아니면 더 나은 방법이 있는지 궁금합니다.

이 테스트는 왜 하려고 하는건지 알 수 있을까요?
랜덤이 정말 잘 동작하는지를 알고싶은 것인가요?

그렇다면 코틀린 팀에서 작성한 이 테스트는 무용지물이..되는 것일까요 ㅠ
https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/test/random/RandomTest.kt

제가 드리고 싶은 말은. 표준 라이브러리에서 제공하는 함수를 테스트 하려는 의도에 대해서입니다.

medAndro replied:
[RaceService.kt](https://github.com/woowacourse/kotlin-racingcar/pull/153/files/b68b3b9b6efc56587d3fa5708bf8edfa62030ea2#diff-20309531194429947a4f0ab16b2381d2f77f3f28c4b10ac091406f96baee83ac) 의 singleRace()가 의도한대로 "0에서 9사이의 랜덤값을 추출해서 전체 리스트의 car 객체들이 전진하거나 그렇지 않는지" 테스트하고 싶었지만 private로 제한되어있어서 어떻게 할까 고민하다가 페어인 밀러가 내부의 동일한 코드를 확률적으로 검사하는 방법을 제안해서 작성하게 되었습니다.

car객체에서 특정한 정수 값을 외부에서 받아 전진을 할지 말지 결정하게 하는 메서드를 만들고 해당 메서드를 테스트하는 방식으로 리팩터링 해볼 예정입니다!

laco-dev commented on src/test/kotlin/ApplicationTest.kt:
```diff
@@ -0,0 +1,156 @@
+import java.io.PrintStream
+import kotlin.random.Random
+
+class ApplicationTest {
```
ApplicationTest에 모든 테스트를 작성 해주셨어요.
만약 이 프로그램에 존재하는 기능이 1000개가 넘는다면 모든 기능을 전부 하나의 테스트 파일에 작성을 해야 할까요?

객체지향 프로그래밍에서 우리는 "객체"를 작성합니다.
객체는 자기만의 역할과 기능을 가지고 있기때문에 객체를 테스트 하게 됩니다.

그렇다면 CarTest를 작성하는게 맞는 방법일지 AppliationTest에 Car 기능 테스트를 모두 작성하는게 맞는지 고민 해보시면 좋겠어요

medAndro replied:
1000개라고 생각해보니 확실히 공감이 가는 말씀이라 생각합니다
객체별로 테스트를 분리해서 특정 객체에 문제가 있는지 파악하기 쉽도록 해보도록 하겠습니다 감사합니다!

laco-dev commented on src/test/kotlin/ApplicationTest.kt:
```diff
@@ -0,0 +1,156 @@
+        }
+    }
+
+    @Test
+    @DisplayName("Int 자료형으로 입력 가능한 1이상의 자연수만 허용한다")
+    fun t3() {
+        setInput("a\n0")
+        assertThrows<IllegalArgumentException> { Application().run() }
+    }
+
+    @Test
+    @DisplayName("Int 자료형으로 입력 가능한 1이상의 자연수만 허용한다")
+    fun t3_1() {
+        setInput("a\n1")
+        assertDoesNotThrow { Application().run() }
+    }
```
지금처럼 실제 프로그램이 동작하는지 까지 테스트로 작성을 할 수도 있지만 테스트 가능한 부분과 테스트 불가능한 부분을 나누는 것이 프로그램을 설계하는데 있어 기본적인 흐름이 됩니다.

그렇다면 우리가 테스트해야 할 기능은 입력이 정상적으로 들어오는지 콘솔로 입력받아본다가 필요한 것일까요?
만약 안드로이드 앱을 만들고 있다면 "앱이 터져야 한다, 앱이 실행되어야 한다"를 테스트 하는 것이 될 것 처럼 보여서요.

앞서 말씀드린 것 처럼 프로그램이 정상적으로 실행되는지 보다는 필요한 역할들을 쉽게 테스트 할 수 있는 단위로 나눠보고 그 기능을 테스트 하는 순서로 이루어져야 합니다.

예를 들어~ 입력을 지금은 콘솔에서 받고 있는데요.
실제 프로그램은 입력이 콘솔이 아닌 키오스크에서 받는다고 가정해보세요.

medAndro replied:
사실 setInput() 내부에서 ByteArrayInputStream()과 setIn()통해 콘솔 입력을 모사하는 방식도 페어를 통해 알게되었고, 현재의 요구사항에선 콘솔에서만 입력받기 때문에 괜찮은 방법이라 생각했었습니다. 

말씀해 주신 것처럼 키오스크 입력과 같은 확장성 있는 코드를 짜기 위해 Application().run()을 통해 코드를 실제 동작시키고 콘솔입력을 모사하는 것이 아닌 입력값을 검증하는 해당 부분에서 테스트할 수 있도록 노력해보겠습니다.

laco-dev commented on src/test/kotlin/ApplicationTest.kt:
```diff
@@ -0,0 +1,156 @@
+        assertThat(raceService.getWinner(cars)).contains("최종 우승자 : a")
+    }
+
+    @Test
+    @DisplayName("전체 기능 테스트")
+    fun t5() {
+        setInput("pobi,woni,jun \n5")
+        val output = setOutput()
+
+        Application().run()
+        assertThat(output.toString().normalize()).contains(
+            """
+            경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).
+            시도할 횟수는 몇 회인가요?
+
+            실행 결과
+            pobi : 
+            """.trimIndent().normalize(),
+        )
+        assertThat(output.toString().normalize()).contains(
+            """
+            
+            최종 우승자 : 
+            """.trimIndent().normalize(),
+        )
+    }
```
이것도 마찬가지로 우리가 테스트 하고 싶은것은 전체 기능 테스트가 아니고, 아래 기능입니다.
> 경주가 끝났을 때 가장 멀리 가 있는 자동차가 우승자다.

laco-dev commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,10 @@
+package racingcar.domain
+
+class Car(
+    val name: String,
```
> 자동차 이름은 5자를 초과할 수 없다.

이 기능은, 자동차의 기능이라고 봐야 할까요? 아니면 자동차를 만드는 것의 기능이라고 봐야 할까요?
아래 코드가 논리적으로 올바른 코드인지 궁금합니다~

```kotlin
val car = Car("Asdfiasjdoiasjdoiadjsaodjasodjasdjsaisajdia", 10)
```

medAndro replied:
현재는 Car 객체를 만들때 마다 매번 입력값이 올바른지 검증하는데, Car내부에서 검사해 예외를 던져주며 비정상적인 이름을 가지는 Car객체가 애초에 생성되지 않게 하는게 합리적일 것 같다고 생각합니다. 수정해보겠습니다!

laco-dev commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,10 @@
+
+class Car(
+    val name: String,
+    var distance: Int = 0,
```
이 코드를 참고해보세요. 어떤 문제가 있을까요?

```kotlin
val car = Car("...")
car.moveForward()
car.moveForward()
car.moveForward()
car.moveForward()
car.moveForward()
car.distance = 0
```

medAndro replied:
자동차 인스턴스의 주행거리를 외부에서 마음대로 조작 가능하다는 문제가 있고
프로퍼티의 setter에 private 접근 제어자를 달아 이동 거리를 임의로 수정할 수 없도록 해결해 보겠습니다

laco-dev commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,10 @@
+    val name: String,
+    var distance: Int = 0,
+) {
+    fun moveForward() = distance++
```
> 전진하는 조건은 0에서 9 사이에서 무작위 값을 구한 후 무작위 값이 4 이상일 경우이다.

앞서 의견과 마찬가지로 "4 이상의 값일 경우 전진한다" 라는 기능은 누구의 기능이라고 봐야 할까요?

> 4 이상의 값이라면 한칸 전진한다.
> 4 미만의 값이라면 정지한다.

laco-dev commented on src/main/kotlin/racingcar/Application.kt:
```diff
@@ -0,0 +1,38 @@
+        print(raceService.getWinner(cars))
+    }
+
+    private fun startGame() {
+        cars.addAll(getCars())
+
+        val raceCount = getRaceCount()
+        println(Messages.GAME_RESULT.message)
+
+        print(raceService.race(raceCount, cars))
+    }
```
```kotlin
val appliation = Application()
application.run()
application.run()
application.run()
application.run()
application.run()
```

이렇게 호출하면 어떤 문제가 있나요??

medAndro replied:
Application() 내부의 cars 리스트가 초기화 되어있지 않아 이전의 경주를 했던 car 인스턴스들이 남게 되는 문제가 발생하고 있고 전체 게임이 끝나면 초기화를 시키도록 개선해 보겠습니다

laco-dev commented on src/main/kotlin/racingcar/InputValidatorService.kt:
```diff
@@ -0,0 +1,56 @@
+import racingcar.domain.Messages
+
+class InputValidatorService {
+    fun getValidatedCarNames(input: String): List<Car> {
+        emptyCheck(input)
+        val names = input.split(",").map { it.trim() }
+        duplicateNameCheck(names)
+
+        names.forEach { name ->
+            nameLengthCheck(name)
+            availableNameCheck(name)
+        }
+        return toCars(names)
+    }
```
InputValidator가 Car를 생성 해도 괜찮을까요?
그렇지 않다면 Car는 어디서 생성하는게 제일 좋을까요?

medAndro replied:
InputValidator는 입력값 검사만 하고 Car는 서비스에서 생성하는게 단일 책임 원칙에 부합할 것 같습니다.

laco-dev commented on src/main/kotlin/racingcar/RaceService.kt:
```diff
@@ -0,0 +1,34 @@
+        return Messages.WINNER.formattedMessage(winners.joinToString(", ") { it.name })
+    }
+
+    fun race(
+        raceCount: Int,
+        cars: List<Car>,
+    ): String {
+        repeat(raceCount) {
+            singleRace(cars)
+            stringBuilder.append("\n")
+        }
+        return stringBuilder.toString()
+    }
```
경주 서비스 라는 비즈니스 모델이 race()라는 함수를 수행했을 때 어떤 결과를 반환하기를 기대할까요?
다른 개발자가 보더라도 아무도 이해를 못할 것 같아요.

medAndro replied:
race() 만으로는 무엇을 하는지 확실히 알기 어려운 것 같습니다 단일 책임 원칙도 위반하는 것 같네요
자동차 경주를 진행하는 부분과 우승자 문자열을 가져오는 부분을 분리하면서 이름을 더 고심하며 지어보도록 하겠습니다.

laco-dev reviewed (changes requested):
만나서 반갑습니다.
자동차 경주 미션동안 잘 부탁드립니다.
여러가지 README와 함수 목록등을 잘 정리해주신 것 같습니다.

아키텍처 관점에서도 어느정도 벌써 관심이 있으신 것 같은데 너무 너무 멀리 나가신 것 같습니다.

우리가 구현하고 집중해야 할 부분은 0부터 10중에 0, 1인데 혼자 8, 9를 고민하고 있는 것 같아요.

그렇다고 0, 1이 잘 구현되었는가? 아닙니다..
과한 욕심과 시도는 리뷰어, 리뷰이 모두의 비용을 크게 증가시키는 문제가 될 수도 있어요.

기본적인 것 부터 현재 구현해야 하는 요구사항에 집중해서 천천히 나아가봅시다 💪🏻

medAndro committed:
refactor: 자동차의 전진 여부를 로직 수정

- 전진 여부를 자동차 객체 내부에서 자동차에게 주어지는 정수값에 의해 결정하도록 변경
- 값의 범위를 넘어가면 예외가 발생하는 코드 추가

medAndro committed:
refactor: OutputView 도입 및 stringBuilder 제거

medAndro committed:
refactor: 자동차 이름을 자동차 객체 내부에서 검사하도록 변경

medAndro committed:
refactor: Car 객체 생성 위치와 컨트롤러 코드 수정

- Car객체를 InputValidator가 아닌 RaceService에서 생성하도록 변경
- 사용자와의 입출력을 콘솔로 받는 View 사용
- run()을 통해 게임 시작시 게임의 초기화를 진행

medAndro committed:
refactor: 테스트 유연성을 위한 의존성 주입 도입

medAndro committed:
fix: 자동차 이름 검증 로직 수정

- 이름 중간에 공백이 존재하는 경우를 허용하도록 수정

medAndro committed:
test: InputValidator클래스의 테스트 코드 추가

medAndro committed:
test: 자동차 모델의 테스트 코드 추가

medAndro committed:
refactor: 테스트 용이성 향상을 위한 인터페이스 적용

medAndro committed:
test: RaceService 클래스의 테스트 코드 추가

medAndro committed:
refactor: 테스트 전용 인터페이스 분리

medAndro commented:
안녕하세요 페로로! 상세한 코드리뷰 남겨주셔서 감사합니다 

리팩터링을 진행하면서 콘솔 입력이 아닌 키오스크에서 입력한다면 어떻게 해야할까를 염두하며 DI와 Interface를 사용해 보았고 Car, InputValidator, RaceService 객체의 메서드를 테스트하는 코드를 작성해보았습니다!

laco-dev commented on src/main/kotlin/racingcar/InputValidator.kt:
```diff
@@ -0,0 +1,51 @@
+import racingcar.domain.Car
+import racingcar.domain.Messages
+
+class InputValidator {
```
도메인 모델 스스로 유효성 검증을 하면 되는데 굳이 Validator를 별도로 빼서 한 이점이 무엇인지 모르겠습니다. 😓 
입력을 통해 생성하는 클래스가 100개라면, InputValidator에는 1000개 이상의 함수를 만들 건가요?

또한 이 컴포넌트는 Console에서 입력받은 것을 검증하는 것 외에도 도메인 로직도 포함하고 있습니다.
2단계에서 UI 로직과 도메인 로직을 분리해야 한다는 것에 이것도 포함되는 주제입니다.

medAndro replied:
Validator를 따로 둬서 도메인의 책임 감소와 재사용성(emptyStringCheck()를 2곳에서 사용하는등)을 고려해서 분리 해봤는데 말씀을 듣고보니 직관성을 해치는 오버엔지니어링인것 같습니다. 2단계에서는 Validator를 제거하고 필요한 객체에서 예외를 던져보도록 리팩터링 해보겠습니다.

laco-dev commented on src/test/kotlin/CarModelTest.kt:
```diff
@@ -0,0 +1,45 @@
+        car = Car("테스트", inputValidator)
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = [4, 5, 6, 7, 8, 9])
+    @DisplayName("랜덤 값에 의한 전진 테스트")
+    fun forwardCarByValue(moveValue: Int) {
+        car.moveByValue(moveValue)
+        assertThat(car.distance).isEqualTo(1)
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = [0, 1, 2, 3])
+    @DisplayName("랜덤 값에 의한 멈춤 테스트")
+    fun stopCarByValue(moveValue: Int) {
+        car.moveByValue(moveValue)
+        assertThat(car.distance).isEqualTo(0)
+    }
```
마찬가지로 자동차 입장에서는 들어오는 숫자가 랜덤인지 아닌지 관심사가 아닙니다.
아래 인터페이스와 구현체의 관계를 보고 어떤 의도를 제시하고 있는지 생각해보시면 좋겠습니다.

```kotlin
interface NumberGenerator {
   fun getNumber(): Int
}

class RandomNumberGenerator: NumberGenerator { ... }
class FixedNumberGenerator: NumberGenerator { ... }
```

medAndro replied:
MoveStrategy 인터페이스를 사용해 Car에 주입하는 방식으로 관심사를 제거해 보겠습니다!

laco-dev commented on src/test/kotlin/RaceServiceTest.kt:
```diff
@@ -0,0 +1,117 @@
+        assertThat(outputView.messages).contains("메다 : -----")
+    }
+
+    @Test
+    @DisplayName("우승자 출력 테스트")
+    fun showWinnerTest() {
+        inputView = TestInputView(listOf("메다,밀러", "10"))
+        raceService = RaceService(random, outputView, inputView, inputValidator)
+
+        assertDoesNotThrow {
+            raceService.carsInitializer()
+            raceService.raceCountInitializer()
+            raceService.doWholeRace()
+            raceService.showCarNamesByWinnerFormat()
+        }
+        val lastMessage = outputView.messages.last()
+        assertThat(lastMessage).isEqualTo("최종 우승자 : 메다, 밀러")
+    }
```
조심스레 다시 의견을 드리지만, 우리는 UI가 어떻게 출력될지를 테스트하는게 아닙니다.
우승자는 "메다", "밀러"라는 차량 2개인 것을 아는 것이지요.

medAndro replied:
RacingGame 도메인에서 우승자 이름들을 반환하도록 리팩터링하고 이 메서드를 테스트 하는것으로 바꿔보겠습니다

laco-dev commented on src/main/kotlin/racingcar/domain/Messages.kt:
```diff
@@ -0,0 +1,21 @@
+package racingcar.domain
+
+enum class Messages(
+    val message: String,
+) {
+    USERINFO_INPUT_CAR_NAME("경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분)."),
+    USERINFO_INPUT_RACE_COUNT("시도할 횟수는 몇 회인가요?"),
+    RACE_RESULT("실행 결과"),
+    WINNER_NAMES_ANNOUNCE("최종 우승자 : %s"),
+
+    ERROR_EMPTY_INPUT("문자열을 입력하지 않으셨습니다."),
+    ERROR_NAME_LENGTH("자동차 이름을 1자 이상 5자 이하로 입력하세요."),
+    ERROR_DUPLICATE_NAME("중복된 자동차 이름이 존재합니다."),
+    ERROR_NOT_AVAILABLE_NAME("지원되지 않는 이름입니다. (한글, 영어, 숫자, 공백 가능)"),
+    ERROR_NOT_POSITIVE("1 이상의 양의 정수(자연수)만 입력이 가능합니다."),
+    ERROR_OVERSIZE_TRY_COUNT("8자리 이하의 자연수만 입력이 가능합니다."),
+    ERROR_MOVE_VALUE("Car의 이동을 결정하는 값은 0~9 이내의 정수만 가능합니다."),
+    ;
+
+    fun formattedMessage(vararg args: String) = String.format(message, *args)
+}
```
Messages는 현재 UI로직과 도메인 로직이 모두 결합되어 있습니다.
UI에서 출력해야할 문구와 도메인에서 스스로 판단할 문구를 모두 분리해주시면 됩니다.

medAndro replied:
각 계층별로 message enum을 따로 분리해보겠습니다...!

laco-dev commented on src/main/kotlin/racingcar/Random.kt:
```diff
@@ -0,0 +1,13 @@
+import racingcar.domain.Configure.Companion.RANDOM_SEED
+import kotlin.random.Random
+
+class Random : RandomGenerator {
+    private val random = Random(RANDOM_SEED)
+
+    override fun nextInt(
+        from: Int,
+        until: Int,
+    ): Int = random.nextInt(from, until)
+}
```
코틀린 표준 함수에서 랜덤을 어떻게 사용하는지 다른 사례들을 한번 찾아보시겠어요?

medAndro replied:
```kt
// 0-9 사이에서 랜덤 정수 추출
Random.nextInt(from = 0, until = 10)
Random.nextInt(10)  
(0..9).random()
(0 until 10).random()

// Int 전체에서 랜덤 정수 추출
Random.nextInt()  
```

IntRange를 사용한 `(0..9).random()`가 직관적인것 같네요...!

laco-dev commented on src/main/kotlin/racingcar/Interfaces.kt:
```diff
@@ -0,0 +1,21 @@
+    ): Int
+}
+
+interface OutputView {
+    fun showMessage(
+        message: String,
+        endWithNewLine: Boolean = false,
+    )
+
+    fun showBlankLine()
+}
+
+interface InputView {
+    fun readInput(): String
+}
```
InputView, OutputView도 인터페이스로서 어떤 것을 입력받을지 출력할지를 기능으로 사용하는 것입니다.
작성하신 구조는 이렇게 만든 것과 차이가 없어 보입니다.

```kotlin
class InputView {
  fun get(): Any
} 

class OutputView {
   fun print(string: String)
}
```

예를 들어 입력 뷰로부터 전화번호를 받아온다거나, 자동차 이름을 받아온다거나, 하는 기능들이 있을 것입니다.
또한 키오스크를 고려하셨다고 하는데, 키오스크에서는 String을 반환할까요? 내부에서는 자동차 모양이 그려진 이미지를 쓴다고 하면, 이걸 String으로 변환해서 주는게 맞을지 의문이 듭니다.

medAndro replied:
View에서 각 입출력 기능별 메서드를 구체적으로 분리시키도록 하겠습니다

laco-dev commented on src/main/kotlin/racingcar/RaceService.kt:
```diff
@@ -0,0 +1,64 @@
+import racingcar.domain.Car
+import racingcar.domain.Messages
+
+class RaceService(
+    private val random: RandomGenerator,
+    private val outputView: OutputView,
+    private val inputView: InputView,
+    private val inputValidator: InputValidator,
+) {
```
제가 보기에는 현재 Application 자체가 컨트롤러나 다름없는데 서비스도 컨트롤러로 보입니다.
서비스는 도메인 로직인가요? 컨트롤러인가요?

서비스 레이어는 무엇이고 어떤 역할을 하는지 다른 개발자도 쉽게 알 수 있을 정도로 저에게 설명해주세요

medAndro replied:
조사해보니 컨트롤러에서 도메인의 비즈니스 로직들을 각각 불러와 사용하는 대신 도메인의 여러 복잡한 상호작용을 하는 비즈니스 로직들을 서비스 레이어에서 캡슐화해서 묶어두고 불러와 컨트롤러가 복잡해지는것을 방지하기 위한 도메인과 컨트롤러 사이에 존재하는것으로 이해했습니다...

굳이 서비스 레이어가 필요 없을것 같아서 2단계에서는 핵심 비즈니스 로직을 도메인에 두는데 집중하고 그것을 컨트롤러에서 바로 불러오는 방식으로 리팩터링 해보도록 하겠습니다.

laco-dev commented on src/main/kotlin/racingcar/RaceService.kt:
```diff
@@ -0,0 +1,64 @@
+    private val cars = mutableListOf<Car>()
+    private var raceCount = 0
+
+    fun carsInitializer() {
+        cars.clear()
+        outputView.showMessage(Messages.USERINFO_INPUT_CAR_NAME.message, true)
+        val carNamesUserInput = inputView.readInput()
+        cars += getCarsByUserInput(carNamesUserInput)
+    }
```
서비스가 생각보다 더 복잡하게 만들어진 것 같아서 의견을 드려봅니다.
결국 우리가 하고 싶은 것은 아래와 같습니다.

1. 자동차들을 준비한다
2. 몇바퀴를 경주할 것인지 판단한다.
3. 자동차를 경주시킨다.
4. 우승자를 찾는다.
5. 우승자의 이름을 출력한다.

참고할만한 슈도 코드를 남겨드릴게요.
직접적으로 어디를 수정하라는 의도가 아닌 단순히 생각을 확장하기 위해서 남기는 내용입니다~

```kotlin
// Controller
val cars: List<Car> = readyCars()
val raceCount: Int = readRaceRound()
val racingGame = RacingGame(cars, ...)

repeat(raceCount) {
   racingGame.race()
   printRoundResult(cars)
}

val winners: List<Car> = racingGame.getWinners()
printGameResult(winners)
```

laco-dev reviewed (approved):
피드백 반영하느라 고생하셨습니다.

MVC를 고려했다고 하지만 실제로 MVC 컴포넌트들에 맞게 구현된 것이 잘 안느껴집니다.
본문에 작성하신 것처럼 테스트가 쉬워졌다는데 제가 보기에는 누구보다도 테스트가 어렵게 느껴집니다...

UI 로직과 도메인 로직을 분리하는 다음 단계를 진행하면서, 컨트롤러에 대한 부분도 명확히 잡아가시면 좋을 것 같습니다.

같이 진행한 페어에게도 드리고 싶은 말은 두 분이서 많이 알고 있다고 생각하여 마음껏 실력을 뽐낸 코드가 실제로는 기본적인 부분도 제대로 알지 못하고 쓴 것 같은 느낌이 듭니다.

요구사항이 너무 쉽게 느껴지시더라도 조금은 마음을 가볍게 잡으시고 기본적인 구현부터 하나씩 단계적으로 나아가면 좋겠습니다.
앞으로 해야할 미션이 산더미처럼 많습니다 ㅎㅎ 😓
